router/pkg/routerplugin/grpc_plugin_client.go

```diff
@@ -16,27 +18,114 @@ type GRPCPluginClient struct {
 
 	pc *plugin.Client
 	cc grpc.ClientConnInterface
+
+	config GRPCPluginClientConfig
+
+	mu sync.RWMutex
+}
+
+type GRPCPluginClientConfig struct {
+	ReconnectTimeout time.Duration
+	PingInterval     time.Duration
+}
+
+var defaultGRPCPluginClientConfig = GRPCPluginClientConfig{
+	ReconnectTimeout: time.Second * 20,
+	PingInterval:     time.Second * 2,
+}
+
+type GRPCPluginClientOption func(*GRPCPluginClientConfig)
+
+func WithReconnectConfig(reconnectTimeout time.Duration, pingInterval time.Duration) GRPCPluginClientOption {
+	return func(c *GRPCPluginClientConfig) {
+		c.ReconnectTimeout = reconnectTimeout
+		c.PingInterval = pingInterval
+	}
 }
 
 var _ grpc.ClientConnInterface = &GRPCPluginClient{}
 
-func newGRPCPluginClient(pc *plugin.Client, cc grpc.ClientConnInterface) (*GRPCPluginClient, error) {
+func newGRPCPluginClient(pc *plugin.Client, cc grpc.ClientConnInterface, options ...GRPCPluginClientOption) (*GRPCPluginClient, error) {
 	if pc == nil || cc == nil {
 		return nil, errors.New("plugin client or grpc client conn is nil")
 	}
 
+	config := defaultGRPCPluginClientConfig
+
+	for _, option := range options {
+		option(&config)
+	}
+
 	return &GRPCPluginClient{
-		pc: pc,
-		cc: cc,
+		pc:     pc,
+		cc:     cc,
+		config: config,
 	}, nil
 }
 
+func (g *GRPCPluginClient) waitForPluginToBeActive() error {
+	timeout := time.After(g.config.ReconnectTimeout)
+	for {
+		select {
+		case <-timeout:
+			return errors.New("plugin was not active in time")
+		default:
+			// we need to use a closure here to avoid race conditions
+			// as we cannot use defer in a for loop
+			shouldContinue, err := func() (bool, error) {
+				g.mu.RLock()
+				defer g.mu.RUnlock()
+				if g.pc == nil {
+					return true, nil
+				}
+
+				clientProtocol, err := g.pc.Client()
+				if err != nil {
+					return false, err
+				}
+
+				if err := clientProtocol.Ping(); err != nil {
+					time.Sleep(g.config.PingInterval)
+					return true, nil
+				}
+
+				return false, nil
+			}()
```

Comment:
Can we factor this out into a function like, say "CheckPluginHealth" or "PingPlugin" ?